models/MonteCarlo.py

```diff
@@ -0,0 +1,142 @@
+A monte carlo model with epsilon greedy policy.
+
+- Episode: A complete run of the model from the start to the end of the training data.
+- Epsilon: The probability of choosing a random action instead of the best action for exploring instead of exploiting.
+- Portfolio Value: init to be 100 at first and then it will be updated based on the reward on each episode to track the value.
+
+'''
+
+class MonteCarloRL(Model):
+    def __init__(self, cleaned_data, actions, n_episodes=1000, epsilon=0.1):
+        super().__init__(state_space=None, action_space=actions)
+        self.cleaned_data = cleaned_data
+        self.actions = actions
+        self.n_episodes = n_episodes
+        self.epsilon = epsilon
+        self.nasdaq_train, self.msci_train, self.nasdaq_test, self.msci_test = self.split_data()
+    
+    @staticmethod
+    def calculate_reward(action, nasdaq_return, msci_return):
+        weight_nasdaq, weight_msci = action
+        portfolio_return = (weight_nasdaq / 100) * nasdaq_return + (weight_msci / 100) * msci_return
+        return portfolio_return
+    
+
+    def split_data(self):
+        train_data = self.cleaned_data[self.cleaned_data['NASDAQ_Date'] < '2020-01-01']
+        test_data = self.cleaned_data[self.cleaned_data['NASDAQ_Date'] >= '2020-01-01']
+
+        nasdaq_train = train_data['NASDAQ_Returns'].values
+        msci_train = train_data['MSCI_Returns'].values
+        nasdaq_test = test_data['NASDAQ_Returns'].values
+        msci_test = test_data['MSCI_Returns'].values
+
+        return nasdaq_train, msci_train, nasdaq_test, msci_test
+
+
+    def train(self, episodes=None):
+        results = []
+        best_action_index = 0
+        n_steps = len(self.nasdaq_train)
+
+        for episode in range(self.n_episodes):
+            portfolio_value = 100  # Initial portfolio value
+            episode_rewards = []
+
+            for t in range(n_steps):
+                if random.uniform(0, 1) < self.epsilon:
+                    action_index = random.randint(0, len(self.actions) - 1)  # Explore
+                else:
+                    action_index = best_action_index  # Exploit
+
+                action = self.actions[action_index]
+     
+                nasdaq_return = self.nasdaq_train[t]
+                msci_return = self.msci_train[t]
+
+                reward = self.calculate_reward(action, nasdaq_return, msci_return)
+                portfolio_value *= (1 + reward / 100)
+                episode_rewards.append(reward)
+
+            avg_reward = np.mean(episode_rewards)
+            if episode == 0 or avg_reward > results[best_action_index]['average_reward']:
+                best_action_index = action_index
+
+            results.append({
+                'episode': episode,
+                'final_portfolio_value': portfolio_value,
+                'average_reward': avg_reward,
+            })
+
+        return pd.DataFrame(results), self.actions[best_action_index]
+    
+    def test(self, optimal_action):
```

Comment:
I think the optimal action here should be evaluated on a state-action pair by state-action pair basis right? it's not a case of finding the singular best action from training and always using that same action during testing, which is what this code seems to be doing